corehq/ex-submodules/pillowtop/pillow/interface.py

```diff
@@ -100,29 +100,94 @@ def run(self):
+                    else:
+                        self.process_with_error_handling(change, context)
                 else:
                     self._update_checkpoint(None, None)
+            self.process_chunk_with_error_handling(changes_chunk, context)
         except PillowtopCheckpointReset:
+            # finish processing any ramining chunk
+            self.process_chunk_with_error_handling(changes_chunk, context)
             self.process_changes(since=self.get_last_checkpoint_sequence(), forever=forever)
 
-    def process_with_error_handling(self, change, context):
+    def process_chunk_with_error_handling(self, changes_chunk, context):
+        """
+        Passes given changes_chunk to the processor for chunked processing
+            If there is an exception in chunked processing, falls back
+            to passing changes one by one to the processor
+        """
+        if not changes_chunk:
+            return
+        failed_changes = set()
+        try:
+            # chunked processing is supported if there is only one processor
+            failed_changes = self.processors[0].process_changes_chunk(self, changes_chunk)
+        except Exception as ex:
+            notify_exception(
+                None,
+                "{pillow_name} Error in processing changes chunk {change_ids}: {ex}".format(
+                    pillow_name=self.get_name(),
+                    change_ids=[c.id for c in changes_chunk],
+                    ex=ex
+                ))
+            # fall back to processing one by one
+            for change in changes_chunk:
+                self.process_with_error_handling(change, context, chunked_fallback=True)
+        else:
+            succeeded_changes = set(changes_chunk) - set(failed_changes)
+            for change in succeeded_changes:
+                self._update_checkpoint(change, context)
+            # fall back to processing one by one for failed changes
+            for change in failed_changes:
+                self.process_with_error_handling(change, context, chunked_fallback=True)
+        self._record_datadog_metrics(changes_chunk, failed_changes)
+
+    def _record_datadog_metrics(self, changes_chunk, failed_changes):
+        tags = ["pillow_name:{}".format(self.get_name())]
+        datadog_counter('commcare.change_feed.chunked.changes', len(changes_chunk), tags=tags)
+        datadog_counter('commcare.change_feed.chunked.changes.exception', len(changes_chunk), tags=tags)
+        datadog_counter('commcare.change_feed.chunked.changes.suceess', len(set(changes_chunk) - set(failed_changes)), tags=tags)
+
+        max_change_lag = (datetime.utcnow() - changes_chunk[0].metadata.publish_timestamp).seconds
+        min_change_lag = (datetime.utcnow() - changes_chunk[-1].metadata.publish_timestamp).seconds
+        perceived_lag = (max_change_lag - min_change_lag) / 2
+        datadog_gauge('commcare.change_feed.chunked.perceived_change_lag', perceived_lag, tags=[
```

Comment:
What does `perceived_change_lag` mean? Is it the amount of time we've been collecting the chunks for?

Similarly I don't quite understand `actual_change_lag` It looks like it's the average of all documents lags, which I'm not sure how it would be used.

How do you think  these will be used?

Reply:
I will make this simple and keep just min/max change-lags. 

Let's say we have 100 changes, with the earliest change 2 seconds ago and last change 1 second ago. If we take 10sec to process these 100 changes, the min_change_lag would be 10sec - 2sec, and max would be 10sec - 1sec.

Also interesting, that in the chunked processing the change lag might actually go up, since we wait to process changes in a minimum chunk. But overall because we process them in chunks, the processing time per change should go down.